test/eoftest/eoftest_runner.cpp

```diff
@@ -0,0 +1,83 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "../utils/utils.hpp"
+#include "eoftest.hpp"
+#include <evmc/evmc.hpp>
+#include <evmone/eof.hpp>
+#include <gtest/gtest.h>
+#include <nlohmann/json.hpp>
+
+namespace json = nlohmann;
+
+namespace
+{
+struct EOFValidationTest
+{
+    struct Case
+    {
+        struct Expectation
+        {
+            evmc_revision rev;
+            bool result;
+            std::string exception;
+        };
+        std::string name;
+        evmc::bytes code;
+        std::vector<Expectation> expectations;
+    };
+    std::unordered_map<std::string, Case> cases;
+};
+
+void from_json(const json::json& j, EOFValidationTest::Case& o)
+{
+    std::optional<evmc::bytes> hex_str{evmc::from_hex(j.at("code").get<std::string>())};
+    if (!hex_str)
+        throw std::invalid_argument{"invalid hex string"};
+    const evmc::bytes code{hex_str.value()};
+    o.code = code;
+
+    for (const auto& [rev, result] : j.at("results").items())
+    {
+        EOFValidationTest::Case::Expectation expectation{};
+        expectation.rev = to_rev(rev);
+        expectation.result = result.at("result").get<bool>();
+        o.expectations.push_back(expectation);
```

Comment:
After you remove `Expectation::exception` this probably can be just
```suggestion
        o.expectations.emplace_back(to_rev(rev), result.at("result").get<bool>());
```